src/public-api-types.ts

```diff
@@ -0,0 +1,126 @@
+/**
+ * Types exposed through the public API
+ *
+ * @note I wanted to use the types in the *-plugin.ts files and just assemble them here,
+ * but the conditional types can't be broken up into separate modules, so I opted to just
+ * write all of them here.
+ */
+import { SpecialInput, FolderInput } from './types'
+
+type NumberInput = number | { value: number; min?: number; max?: number }
+
+type Point2DInput = [NumberInput, NumberInput] | { x: NumberInput; y: NumberInput }
+type Point3DInput = [NumberInput, NumberInput, NumberInput] | { x: NumberInput; y: NumberInput; z: NumberInput }
+
+type IntervalInput = {
+  value: [number, number]
+  min: number
+  max: number
+}
+
+type ImageInput = {
+  image: undefined
+}
+
+type SelectInput = { options: any[] | Record<string, any>; value?: any }
+
+type ColorInput = {
+  r: number
+  g: number
+  b: number
+  a?: number
+}
+
+type SpringInput = {
+  tension: number
+  friction: number
+  mass?: number
+}
+
+type BooleanInput = boolean
+
+type StringInput = string
+
+type SchemaItem =
+  | NumberInput
+  | IntervalInput
+  | SpecialInput
+  | Point2DInput
+  | Point3DInput
+  | ImageInput
+  | SelectInput
+  | ColorInput
+  | SpringInput
+  | BooleanInput
+  | StringInput
+  | FolderInput<any>
+
+type NotAPrimitiveType = { ____: 'NotAPrimitiveType' }
+
+type PrimitiveToValue<S> = S extends { r: number; g: number; b: number; a: number }
+  ? { r: number; g: number; b: number; a: number }
+  : S extends { r: number; g: number; b: number }
+  ? { r: number; g: number; b: number }
```

Comment:
Naive question, is there a way to simplify this using the `ColorInput` type?

Reply:
Sure, fixed it in 173f37a.